driver-async/src/main/com/mongodb/async/client/gridfs/GridFS.java

```diff
@@ -0,0 +1,205 @@
+                    fileCollection.createIndex(
+                            new Document("filename", 1).append("uploadDate", 1), new SingleResultCallback<Void>() {
+
+                        @Override
+                        public void onResult(final Void result, final Throwable t) {
+                            //TODO log
+                            if (t != null) {
+                                //failed to create index
+                            } else {
+                                // index ok
+                            }
+                        }
+                    });
+                }
+            }
+        });
+
+        chunkCollection.count(new SingleResultCallback<Long>() {
+            @Override
+            public void onResult(final Long result, final Throwable t) {
+                if (t != null) {
+                    //TODO log error
+                } else if (result < 1000) {
+                    chunkCollection.createIndex(new Document("files_id", 1).append("n", 1),
+                            new CreateIndexOptions().unique(true), new SingleResultCallback<Void>() {
+
+                        @Override
+                        public void onResult(final Void result, final Throwable t) {
+                            //TODO log
+                            if (t != null) {
+                                //failed to create index
+                            } else {
+                                // index ok
+                            }
+                        }
+                    });
+                }
+            }
+        });
+    }
+
+    /**
+     * Instantiate a new GridFSFile.
+     *
+     * @return GridFSFile
+     */
+    public GridFSFile create() {
+        GridFSFile gridFSFile = new GridFSFile(chunkSize);
+        gridFSFile.setId(new ObjectId());
+        gridFSFile.setUploadDate(new Date());
+        return injectContext(gridFSFile);
+    }
+
+    /**
+     * Find a GridFSFile by its id.
+     *
+     * @param id the id as ObjectId
+     * @param callback callback that is completed when the file is available
+     */
+    public void find(final ObjectId id, final SingleResultCallback<GridFSFile> callback) {
+        fileCollection.find(new Document("_id", id)).first(new SingleResultCallback<GridFSFile>() {
+            @Override
+            public void onResult(final GridFSFile result, final Throwable t) {
+                if (t != null) {
+                    callback.onResult(null, t);
+                } else {
+                    callback.onResult(injectContext(result), null);
+                }
+            }
+        });
+    }
+
+    /**
+     * List all GridFSFiles, restricted by the provided query.
+     * @param query a document to limit the results that are returned
+     * @param callback callback that will be completed when the result is available
+     */
+    public void listFiles(final Document query, final SingleResultCallback<List<GridFSFile>> callback) {
```

Comment:
Is this what you have in mind?

````java
    public MongoIterable<GridFSFile> find(final Object query) {
        return fileCollection.find(query).sort(new Document("filename", 1));
    }
````

This brings up one of the things I wanted to discuss:
The GridFSFile class needs a reference to the MongoCollections it uses. I am currently using a similar workaround as the sync driver ( by calling the injectContext() method after the GridFSFile has been retrieved, which sets the required references).
I think it would be a lot nicer, if we could access the MongoCollection (and MongoDatabase) a Document has been retrieved from through the DecoderContext in the Codec. That way these fields could be initialized when the Object is created in the Codec. However, I am currently not sure if this is something the DecoderContext is meant for or if this has any other impact.